common/remote/sockfile.cpp

```diff
@@ -3792,53 +3915,89 @@ static IOutputMetaData *getTypeInfoOutputMetaData(IPropertyTree &actNode, const
     }
 }
 
-class CRemoteDiskReadActivity : public CSimpleInterfaceOf<IRemoteActivity>, implements IVirtualFieldCallback
+class CRemoteDiskBaseActivity : public CSimpleInterfaceOf<IRemoteActivity>, implements IVirtualFieldCallback
 {
+protected:
     StringAttr fileName;
+    Linked<IOutputMetaData> outMeta;
+    unsigned __int64 chooseN = 0;
+    unsigned __int64 limit = 0;
+    unsigned __int64 processed = 0;
+    Owned<const IDynamicTransform> translator;
+    bool outputGrouped = false;
+    bool opened = false;
+    bool eofSeen = false;
+    bool canMatchAny = false;
+    const RtlRecord *record = nullptr;
+
+public:
+    CRemoteDiskBaseActivity(IHThorCompoundBaseArg *helper, IOutputMetaData *expectedMeta, IOutputMetaData *projectedMeta)
+    {
+        outMeta.set(helper->queryOutputMeta());
+        canMatchAny = helper->canMatchAny();
+        if (!canMatchAny)
+            eofSeen = true;
+        record = &expectedMeta->queryRecordAccessor(true);
+        translator.setown(createRecordTranslator(projectedMeta->queryRecordAccessor(true), *record));
+    }
+// IRemoteActivity impl.
+    virtual unsigned __int64 queryProcessed() const override
+    {
+        return processed;
+    }
+    virtual IOutputMetaData *queryOutputMeta() const override
+    {
+        return outMeta;
+    }
+    virtual bool isGrouped() const override
+    {
+        return outputGrouped;
+    }
+//interface IVirtualFieldCallback
+    virtual const char * queryLogicalFilename(const void * row) override
+    {
+        const char * filename = nullptr; // MORE: This needs to be passed from the client to dafilesrv
+        return filename ? filename : "";
+    }
+};
+
+class CRemoteDiskReadActivity : public CRemoteDiskBaseActivity
+{
+    typedef CRemoteDiskBaseActivity PARENT;
+
     Linked<IHThorDiskReadArg> helper;
     CThorContiguousRowBuffer prefetchBuffer;
-    IArrayOf<IKeySegmentMonitor> segMonitors;
     Owned<ISourceRowPrefetcher> prefetcher;
     Owned<ISerialStream> inputStream;
     Owned<IFileIO> iFileIO;
-    Linked<IOutputMetaData> outMeta;
-    Owned<const IDynamicTransform> translator;
-    unsigned __int64 chooseN = 0;
-    unsigned __int64 limit = 0;
-    unsigned __int64 processed = 0;
     unsigned __int64 startPos = 0;
     bool compressed = false;
     bool inputGrouped = false;
-    bool outputGrouped = false;
-    bool opened = false;
-    bool eofSeen = false;
     bool cursorDirty = false;
-    bool canMatchAny = false;
     mutable bool eogPending = false;
     mutable bool someInGroup = false;
-    const RtlRecord *record = nullptr;
     RtlDynRow *filterRow = nullptr;
     RowFilter actualFilter;
 
     void checkOpen()
     {
-        if (!canMatchAny)
```

Comment:
I think this now means the file will be opened before canMatchAny is checked.  Is that what you wanted?